### PR TITLE
Category list: sorting

### PR DESCRIPTION
**Pull request by florv**
- Button to sort alphabetically
- Sort by drag and drop

### CODE REVIEW

File: indico/htdocs/js/indico/modules/category_management.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                return $(a).data('category-title').localeCompare($(b).data('category-title'));
+            });
+            $.each(sortedRows, function() {
+                $(this).detach().appendTo($tbody);
+            });
+        }
+
+        function setOrderAjax(order) {
+            return $.ajax({
+                url: $table.data('sort-href'),
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: `sort-url`?

File: indico/modules/categories/controllers/management.py [hunk 2]
```diff
@@ -34,3 +35,15 @@ def _category_query_options(self):
...
+
+class RHManageSubcategories(RHManageCategoryBase):
+    def _checkParams(self):
+        RHManageCategoryBase._checkParams(self)
+        self.subcategories = Category.find().filter(Category.id.in_(request.json['categories'])).all()
+        self.subcategories_ids = request.json['categories']
+
+    def _process(self):
+        subcategories = {category.id: category for category in self.subcategories}
+        for (position, id_) in enumerate(self.subcategories_ids):
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: usually we don't use () for tuple unpacking

File: indico/htdocs/js/indico/modules/category_management.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                actionCallback: undo,
+                feedbackMessage: $T.gettext("The category list order has been restored."),
+                duration: 10000,
+                message: $T.gettext("The category list has been sorted.")
+            });
+        });
+
+        $tbody.sortable({
+            axis: 'y',
+            containment: 'parent',
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: can't move stuff above the first or below the last item depending on the cursor position on the element

File: indico/htdocs/js/indico/modules/category_management.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+(function() {
+    'use strict';
+
+    $(document).ready(function() {
+        setupCategoriesTable();
+    });
+
+    function setupCategoriesTable() {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: `setupCategoryTable`

File: indico/htdocs/js/indico/modules/category_management.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+(function() {
+    'use strict';
+
+    $(document).ready(function() {
+        setupCategoriesTable();
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: call from script tag in the template

File: indico/modules/categories/controllers/management.py [hunk 6]
```diff
@@ -34,3 +35,15 @@ def _category_query_options(self):
     def _process(self):
         return WPCategoryManagement.render_template('management/content.html', self.category, 'content',
                                                     categories=self.category.children)
+
+
+class RHManageSubcategories(RHManageCategoryBase):
+    def _checkParams(self):
+        RHManageCategoryBase._checkParams(self)
+        self.subcategories = Category.find().filter(Category.id.in_(request.json['categories'])).all()
+        self.subcategories_ids = request.json['categories']
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: `subcategory_ids`

File: indico/modules/categories/controllers/management.py [hunk 7]
```diff
@@ -34,3 +35,15 @@ def _category_query_options(self):
     def _process(self):
         return WPCategoryManagement.render_template('management/content.html', self.category, 'content',
                                                     categories=self.category.children)
+
+
+class RHManageSubcategories(RHManageCategoryBase):
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: `RHSortSubcategories`?

File: indico/modules/categories/controllers/management.py [hunk 8]
```diff
@@ -34,3 +35,15 @@ def _category_query_options(self):
...
+    def _checkParams(self):
+        RHManageCategoryBase._checkParams(self)
+        self.subcategories = Category.find().filter(Category.id.in_(request.json['categories'])).all()
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: :warning: no access checks!

File: indico/modules/categories/controllers/management.py [hunk 9]
```diff
@@ -34,3 +35,15 @@ def _category_query_options(self):
...
+        self.subcategories_ids = request.json['categories']
+
+    def _process(self):
+        subcategories = {category.id: category for category in self.subcategories}
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: subcategories = {c.id: c for c in self.category.children}

File: indico/htdocs/js/indico/modules/category_management.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    });
+
+    function setupCategoriesTable() {
+        var $table = $('table.categories-management');
+        var $tbody = $table.find('tbody');
+        var $categoryRowSelector = 'tr[data-category-id]';
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: I'm not a fan of `$` prefixes but for this one it's wrong for sure since it's not even a jQuery object but just a string

File: indico/htdocs/js/indico/modules/category_management.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        function restoreCategoriesOrder(order) {
+            $.each(order, function(index, id) {
+                $tbody.find('[data-category-id=' + id + ']').detach().appendTo($tbody);
+            });
+        }
+
+        function sortCategories() {
+            var sortedRows = $tbody.find($categoryRowSelector).sort(function(a, b) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: you can simplify this a bit:

    $tbody.find($categoryRowSelector).detach().sort(...).appendTo($tbody);

File: indico/modules/categories/controllers/management.py (hunk 8 above)
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: btw, `find()` without arguments doesn't make much sense. you'd use the `query` property in that case. the only purpose of `find()` is to provide some sugar (it accepts *args and **kwargs like filter/filter_by and some special ones). but even though i added it initially i'm more and more tempted towards using `.query.filter(..)` and `.query.filter_by()` in favor of `.find()` simply because it's clearer and more like standard sqlalchemy

File: indico/htdocs/js/indico/modules/category_management.js (hunk 3 above)
Comment (florv) [inline indico/htdocs/js/indico/modules/category_management.js]: I've added `tolerance: 'pointer'`which makes the drop position being computed from the position of the mouse instead of the position of the dragged element.

File: indico/htdocs/js/indico/modules/category_management.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+(function(global) {
+    'use strict';
+
+    global.setupCategoryTable = function() {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: `global.setupCategoryTable  = function setupCategoryTable() {`

that way tracebacks etc are nicer

File: indico/htdocs/js/indico/modules/category_management.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            cursor: 'move',
+            handle: '.js-handle',
+            items: '> tr',
+            tolerance: 'pointer',
+            update: function() {
+                setOrderAjax(getCategoriesOrder());
+            }
+        });
+
+        function getCategoriesOrder() {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: `getCategoryOrder`

File: indico/htdocs/js/indico/modules/category_management.js [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            return $.map($tbody.find(categoryRowSelector), function(row) {
+                return $(row).data('category-id');
+            });
+        }
+
+        function restoreCategoriesOrder(order) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: `restoreCategoryOrder`

File: indico/htdocs/js/indico/modules/category_management.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            handle: '.js-handle',
+            items: '> tr',
+            tolerance: 'pointer',
+            update: function() {
+                setOrderAjax(getCategoriesOrder());
+            }
+        });
+
+        function getCategoriesOrder() {
+            return $.map($tbody.find(categoryRowSelector), function(row) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/category_management.js]: return $tbody.find(categoryRowSelector).map(...).toArray();

File: indico/modules/categories/blueprint.py [hunk 16]
```diff
@@ -26,6 +26,7 @@
 
 # Management
 _bp.add_url_rule('/manage/', 'manage_content', RHManageCategoryContent)
+_bp.add_url_rule('/manage/subcategories', 'subcategories', RHSortSubcategories, methods=('POST',))
```
Comment (ThiefMaster) [inline indico/modules/categories/blueprint.py]: `'/manage/sort-subcategories', 'sort_subcategories'` or
`'/manage/subcategories/sort', 'sort_subcategories'`?

File: indico/modules/categories/controllers/management.py [hunk 17]
```diff
@@ -34,3 +35,14 @@ def _category_query_options(self):
     def _process(self):
         return WPCategoryManagement.render_template('management/content.html', self.category, 'content',
                                                     categories=self.category.children)
+
+
+class RHSortSubcategories(RHManageCategoryBase):
+    def _checkParams(self):
+        RHManageCategoryBase._checkParams(self)
+        self.subcategory_ids = request.json['categories']
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: I'd just do that in `_process` - no need to put all request data on `self` unless it's useful for inheritance etc

File: indico/htdocs/js/indico/modules/category_management.js (hunk 13 above)
Comment (florv) [inline indico/htdocs/js/indico/modules/category_management.js]: Shouldn't it be plural since it performs the operation on more than one category?
IMHO with the singular name one expect to have to pass the category on which to perform the operation in argument.

(This comment applies to the similar one bellow too.)

File: indico/modules/categories/blueprint.py (hunk 16 above)
Comment (florv) [inline indico/modules/categories/blueprint.py]: I prefer `/manage/sort-subcategories` since `subcategories/` would unexpectedly 404.
